update readme.md (#212)
wormhole-foundation/example-native-token-transfers: evm/README.md

# Message Lifecycle (EVM)

1. **Send**

> A client calls on [`transfer`] to initiate an NTT transfer. The client must specify at minimum, the amount of the transfer, the recipient chain, and the recipient address on the recipient chain. [`transfer`] also supports a flag to specify whether the NttManager should queue transfers if they are rate-limited (see [link to Rate Limiting Section]). Clients can also include additional instructions to forward along to its peer NTT Manager on the recipient chain to execute. Depending on mode set in the initial configuration of the NttManager contract, transfers are either "locked" or "burned". Once the transfer has been forwarded to the Transceiver, the NttManager emits the `TransferSent` event.

Reviewer: ```suggestion
A client calls [`transfer`] to initiate an NTT transfer. The client must specify, at minimum, the amount of the transfer, the recipient chain, and the recipient address on the recipient chain. [`transfer`] also supports a flag to specify whether the NttManager should queue or revert transfers if they are rate-limited (see [link to Rate Limiting Section]). Depending on mode set in the initial configuration of the NttManager contract, transfers are either "locked" or "burned". Once the transfer has been forwarded to the Transceiver, the NttManager emits the `TransferSent` event.
```
there's no additional information that goes to the ntt manager peer, right?

Author: yea its to the Transceiver - will update

_Events_
``` solidity
/// @notice Emitted when a message is sent from the nttManager.
/// @dev Topic0
///      0x9716fe52fe4e02cf924ae28f19f5748ef59877c6496041b986fbad3dae6a8ecf
/// @param recipient The recipient of the message.
/// @param amount The amount transferred.
/// @param fee The amount of ether sent along with the tx to cover the delivery fee.
/// @param recipientChain The chain ID of the recipient.
/// @param msgSequence The unique sequence ID of the message.
event TransferSent(
    bytes32 recipient, uint256 amount, uint256 fee, uint16 recipientChain, uint64 msgSequence
);
```

2. **Rate Limit**

> A transfers can be rate-limited (see [here](../README.md#rate-limiting-and-cancel-flows) for more details), both on the source and destination chains. If a transfer is rate-limited on the source chain and the `shouldQueue` flag is enabled, it is added to an outbound queue. The transfer can be released after the configured `_rateLimitDuration` has expired via the [`completeOutboundQueuedTransfer`] method. The `OutboundTransferQueued` and `OutboundTransferRateLimited` events are emitted.

Reviewer 2: "A transfers" should be "A transfer"

> During the transfer process, transfers can be rate-limited (see the Rate-Limiting and Cancel Flows section in the main README.md for more details), both on the source and destination chains. If a transfer is rate-limited on the source chain and the `shouldQueue` flag is enabled, they are added into an outbound queue. The transfer can be released after the configured `_rateLimitDuration` has expired via the [`completeOutboundQueuedTransfer`] method. The `OutboundTransferQueued` and `OutboundTransferRateLimited` events are emitted.

Reviewer: it should be possible to hyperlink the md section in the main readme

If the client attempts to release the transfer from the queue before the expiry of the `rateLimitDuration`, the contract reverts with a `OutboundQueuedTransferStillQueued` error.

> Similarly, tranfers that are rate-limited on the destination chain are added to an inbound queue. These transfers can be released from the queue via the [`completeInboundQueuedTransfer`] method. The `InboundTransferQueued` event is emitted.

Reviewer 2: "transfers" is spelled incorrectly

_Events_
``` solidity
/// @notice Emitted whenn an outbound transfer is queued.
/// @dev Topic0
///      0x69add1952a6a6b9cb86f04d05f0cb605cbb469a50ae916139d34495a9991481f.
/// @param queueSequence The location of the transfer in the queue.
event OutboundTransferQueued(uint64 queueSequence);
```

``` solidity
/// @notice Emitted when an outbound transfer is rate limited.
/// @dev Topic0
///      0x754d657d1363ee47d967b415652b739bfe96d5729ccf2f26625dcdbc147db68b.
/// @param sender The initial sender of the transfer.
/// @param amount The amount to be transferred.
/// @param currentCapacity The capacity left for transfers within the 24-hour window.
event OutboundTransferRateLimited(
    address indexed sender, uint64 sequence, uint256 amount, uint256 currentCapacity
);
```

``` solidity
/// @notice Emitted when an inbound transfer is queued
/// @dev Topic0
///      0x7f63c9251d82a933210c2b6d0b0f116252c3c116788120e64e8e8215df6f3162.
/// @param digest The digest of the message.
event InboundTransferQueued(bytes32 digest);
```

3. **Transmit**

> Once the NttManager forwards the message to the Transceiver, the message is transmitted via the [`sendMessage`] method. The method signature if enforced by the [`Transceiver`] but transceivers are free to determine their own implementation for transmitting messages. (e.g A message routed through the Wormhole Transceiver can be sent via automatic relaying (AR), via a specialized or custom relayer, or manually).

Reviewer 2: "if" should be "is" after "The method signature"

Once the message has been transmitted, the contract emits the `SendTransceiverMessage` event.

_Events_
``` solidity
/// @notice Emitted when a message is sent from the transceiver.
/// @dev Topic0
///      0x53b3e029c5ead7bffc739118953883859d30b1aaa086e0dca4d0a1c99cd9c3f5.
/// @param recipientChain The chain ID of the recipient.
/// @param message The message.
event SendTransceiverMessage(
    uint16 recipientChain, TransceiverStructs.TransceiverMessage message
);
```

4. **Receive**

> Once a message has been transmitted across the wire, an off-chain process (e.g. a relayer) will forward the message to the corresponding Transceiver on the recipient chain. The relayer interacts with the transceiver via an entrypoint for receiving messages For example, the client will call the [`receiveWormholeMessage`] method on the WormholeTransceiver contract Wormhole to execute the message. Once the `ReceiveMessage` event is emitted during this process.

Reviewer: ```suggestion
Once a message has been emitted by a Transceiver on the sending, an off-chain process (e.g. a relayer) will forward the message to the corresponding Transceiver on the recipient chain. The relayer interacts with the transceiver via an entrypoint for receiving messages For example, the client will call the [`receiveWormholeMessage`] method on the WormholeTransceiver contract Wormhole to execute the message. Once the `ReceiveMessage` event is emitted during this process.
```

This method should also forward the message to the NttManager on the recipient chain.

> Note that the the Transceiver interface does not enforce the method signature abstractly because receiving messages may be specific to the way in which a transceiver consumes messages.

Reviewer: ```suggestion
Note that the the Transceiver interface does not declare a signature for this method because receiving messages is specific to each transceiver, and a one size fits all solution would be overly restrictive.
```

The NttManager contract allows an M of N threshold for attestations to determine whether a message can be safely executed. For example, if the threshold requirement is 1, the message will be executed after a single Transceiver delivers a valid attestation. If the threshold requirement is 2, the message will only be executed after two Transceivers deliver valid attestations. When a message is attested to by a Transceiver, the contract emits the `MessageAttestedTo` event.

> NTT implements replay protection, so if a Transceiver attempts to deliver a message attestation twice, the contract reverts with `TransceiverAlreadyAttestedToMessage` error. NTT also implements replay protection against reexecuting messages. This check also acts as reentrancy protecion as well. If a message had already been executed, the contract ends execution early and emits the `MessageAlreadyExecuted` event instead of reverting via an error. This mitigates the possibility of race conditions from transceivers attempting to deliver the same message when (threshold < number of transceiver messages and notifies the client (off-chain process) so they don't attempt redundant message delivery

Reviewer 2: Missing period at the end.
